File: 0061-rotate-list/0061-rotate-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def rotateRight(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        length = 0
        tempHead = head
        while tempHead:
            length += 1
            tempHead = tempHead.next
        if length == 0:
            return head
        temp = k % length
        if temp == 0:
            return head
        dummy = ListNode()
        curr = dummy
        while length > temp:
            curr.next = head
            curr = curr.next
            head = head.next
            length -= 1
        curr.next = None
        tempHead = head
        while tempHead:
            if tempHead.next == None:
                tempHead.next = dummy.next
                break
            tempHead = tempHead.next
        return head
                
        

File: 0061-rotate-list/0061-rotate-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* rotateRight(ListNode* head, int k) {
        int length = 0;
        ListNode* temp = head;
        while(temp){
            temp = temp->next;
            length ++;
        }
        if(length == 0){
            return head;
        }
        k = k % length;
        if(k == 0){
            return head;
        }
        temp = head;
        ListNode* dummy = new ListNode();
        ListNode* curr = dummy;
        while(length > k){
            curr->next = temp;
            curr = curr->next;
            temp = temp->next;
            length --;
        }
        curr->next = nullptr;
        head = temp;
        
        while(temp){
            if(temp->next == nullptr){
                temp->next = dummy->next;
                break;
            }
            temp = temp->next;
        }
        return head;
    }
};